### PR TITLE
AST separation

### PR DESCRIPTION
**Pull request by performantdata**
This involves separating the AST code into its own published module. Deprecation of what we throw away or change in the API should happen in a new 2.x branch, if desired.

This PR will help solve #461 and #466.

### CODE REVIEW

**Comment by performantdata** [comment]
I based this branch on #724.

File: modules/ast/src/main/scala/sangria/ast/SourceMapper.scala [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package sangria.parser
+package sangria.ast
```
Comment (yanns) [inline modules/ast/src/main/scala/sangria/ast/SourceMapper.scala:1]: Maybe we can revert this change? This would result in much less changes.

File: modules/benchmarks/src/main/scala/sangria/benchmarks/OverlappingFieldsCanBeMergedBenchmark.scala [hunk 2]
```diff
@@ -119,7 +116,7 @@ class OverlappingFieldsCanBeMergedBenchmark {
                |     xingId {
              """.stripMargin)
 
-    for (i <- 1 to size)
+    for (_ <- 1 to size)
```
Comment (yanns) [inline modules/benchmarks/src/main/scala/sangria/benchmarks/OverlappingFieldsCanBeMergedBenchmark.scala:119]: It'd help the review if you can keep the diff as minimal as possible.

**Comment by yanns** [comment]
I'll try to use a snapshot later to check how important the breaking changes are.

File: build.sbt [hunk 3]
```diff
@@ -178,9 +308,23 @@ lazy val root = project
...
   .disablePlugins(MimaPlugin)
 
+lazy val ast = project.in(file("modules/ast"))
+  .withId("sangria-ast")
+  .settings(scalacSettings ++ shellSettings)
+  .settings(
+    name := "sangria-ast",
+    description := "Scala GraphQL AST representation",
+
+    //FIXME This really shouldn't depend on any other libraries.
```
Comment (yanns) [inline build.sbt:318]: Avoid FIXME in the code that are never fixed ;)
Better adding an issue and linking the issue here.

File: modules/ast/src/main/scala/sangria/ast/SourceMapper.scala (hunk 1 above)
Comment (performantdata) [reply modules/ast/src/main/scala/sangria/ast/SourceMapper.scala:1]: The `SourceMapper` belongs with the `ast` library. Well, not all of it. The `render` methods don't belong there. But the identification of the source and the content of the source belong there.

But yes, this is the main change. It was intentional.

File: build.sbt (hunk 3 above)
Comment (performantdata) [reply build.sbt:318]: Not sure what you mean. A "FIXME" will show up in the IDE list of problems. And the solution to this one is removing the `render` methods, which…will get done eventually? next, even? That's part of the discussion.

There are a lot of issues that don't have GitHub issues tracking them. And I don't even know whether they're issues, because it's hard to have that discussion with you guys, over Gitter.

File: modules/ast/src/main/scala/sangria/ast/SourceMapper.scala (hunk 1 above)
Comment (performantdata) [reply modules/ast/src/main/scala/sangria/ast/SourceMapper.scala:1]: My _belief_ is that it isn't disruptive. I doubt that users create their own `SourceMapper`s, or use the `SourceMapper`s directly. The only reason that i could see for doing so would be internationalization of the error messages, and then there are a lot more such messages in the `Violation`s.

File: build.sbt (hunk 3 above)
Comment (yanns) [reply build.sbt:318]: In my experience, FIXME in the code stay forever and are never tackled. That's why I try to avoid them now.
But having that as a github issue will not ensure that it will be tackle. ;) But at least it's more visible to eventual contributors.

Comment (performantdata) [reply build.sbt:318]: You realize of course that you have six-year-old GitHub issues in the project. 😏 And the build spits out 100+ warnings, with more warnings from IntelliJ. To me, a "FIXME" or "TODO" is just another tool in the drawer. It's useful when the issue doesn't warrant a full-blown GitHub issue. The IDE tracks them, in the Problems tab.

This one in particular I expect to be addressed on the way to separating out the localized string from the Sangria core code. _That_ warrants a GH issue, IMO.

**Review by yanns — COMMENTED** [review]
I've opened a PR with some suggestions: https://github.com/performantdata/sangria/pull/3

File: modules/core/src/main/scala/sangria/renderer/QueryRenderer.scala [hunk 4]
```diff
@@ -485,6 +488,13 @@ object QueryRenderer {
   def render(node: AstNode, config: QueryRendererConfig = Pretty, indentLevel: Int = 0): String =
     renderNode(node, config, Indent(config, indentLevel, indentLevel))
 
+  def renderPretty(node: AstNode): String = node match {
+    case _: Value => render(node, PrettyInput)
+    case _ => render(node, Pretty)
+  }
+
+  def renderCompact(node: AstNode): String = render(node, Compact)
```
Comment (yanns) [inline modules/core/src/main/scala/sangria/renderer/QueryRenderer.scala:496]: This would be something to backport to the 2.x version before, deprecating the `.renderXXX` methods on the AST. This would be a migration path to smoothly transition to this,

Comment (performantdata) [reply modules/core/src/main/scala/sangria/renderer/QueryRenderer.scala:496]: Yep. I have [ideas of what to deprecate](https://github.com/performantdata/sangria/discussions/1#discussioncomment-1573890). Just need to see what gets committed.

**Comment by performantdata** [comment]
https://github.com/performantdata/sangria/pull/3 is cherry-picked.

File: modules/core/src/main/scala/sangria/macros/AstLiftable.scala [hunk 5]
```diff
@@ -142,12 +142,12 @@ trait AstLiftable {
   }
 
   implicit def liftDocument: Liftable[Document] = Liftable { case doc @ Document(d, c, p, _) =>
-    q"_root_.sangria.ast.Document($d, $c, $p, _root_.scala.Some(new _root_.sangria.ast.DefaultSourceMapper(${doc.sourceMapper.get.id}, _root_.org.parboiled2.ParserInput(${doc.source.get}))))"
+    q"_root_.sangria.ast.Document($d, $c, $p, _root_.scala.Some(new _root_.sangria.parser.DefaultSourceMapper(${doc.sourceMapper.get.id}, _root_.org.parboiled2.ParserInput(${doc.source.get}))))"
```
Comment (yanns) [inline modules/core/src/main/scala/sangria/macros/AstLiftable.scala]: What I forgot is that we need to use `SourceMapperInput` here also.
I'm split between having a helper method converting `org.parboiled2.ParserInput` to `SourceMapperInput`.
Or making `SourceMapperInput` using implicit to find the right instance so that we "only" need to import the right package.

Comment (yanns) [reply modules/core/src/main/scala/sangria/macros/AstLiftable.scala]: I see now that you did that here: https://github.com/sangria-graphql/sangria/pull/777/commits/ed755de5dfe3e4d286a18a4b921b965a9b55d0c9
👍

File: modules/core/src/main/scala/sangria/package.scala [hunk 6]
```diff
@@ -0,0 +1,15 @@
+/** Sangria GraphQL server library.
+  *
+  * Sangria is a library that provides parsing, validation, execution and other services for GraphQL
+  * queries.
+  *
+  * It typically requires other libraries to build a complete GraphQL service—including, perhaps,
+  * ones that provide a service interface (e.g.,
+  * [[https://doc.akka.io/docs/akka-http/current/index.html Akka HTTP]]) and a database interface
+  * (e.g., [[https://scala-slick.org/ Slick]])—as well as custom code to bridge the gap between data
+  * representations that are natural to the database vs. to the GraphQL schema.
```
Comment (yanns) [inline modules/core/src/main/scala/sangria/package.scala:10]: My experience is that new users read the website and not the scaladoc.
So I'd prefer not linking to any other libraries here, keeping the sources neutral.

```suggestion
  * It typically requires other libraries to build a complete GraphQL service
  * —including, perhaps, ones that provide a HTTP service interface and a database interface—
  * as well as custom code to bridge the gap between data representations that are natural
  * to the database vs. to the GraphQL schema.
```

File: modules/core/src/main/scala/sangria/parser/package.scala [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package sangria
+
+package object parser {
+
+  @deprecated("please use `sangria.ast.SourceMapper`", "3.0.0")
+  type SourceMapper = sangria.ast.SourceMapper
```
Comment (yanns) [inline modules/core/src/main/scala/sangria/parser/package.scala:6]: Should we add a small test to make sure we provide this backwards compatibility?
During a migration from 2.x to 3.x, this is very important to avoid compilation errors.

**Review by yanns — COMMENTED** [review]
This is looking better & better! 👍

File: modules/core/src/main/scala/sangria/parser/package.scala (hunk 7 above)
Comment (performantdata) [reply modules/core/src/main/scala/sangria/parser/package.scala:6]: What are you thinking? What would the test be?

Comment (yanns) [reply modules/core/src/main/scala/sangria/parser/package.scala:6]: Very simple, just using `sangria.parser.SourceMapper`.
```
val mapper: sangria.parser.SourceMapper = new DefaultSourceMapper...
succeeded
```

File: modules/core/src/main/scala/sangria/package.scala (hunk 6 above)
Comment (performantdata) [reply modules/core/src/main/scala/sangria/package.scala:10]: I also recognize that new users use the website docs first. I don't see how that's relevant. Yes, this scaladoc is for the users who are digging for a deeper understanding.

I felt that pointing to a specific example of "an HTTP service interface" was clearer than using a term ("HTTP service interface") that may be ambiguous to the reader. I picked the ones that I thought most users would know. My text first read like yours, without the "(e.g., …)"s.

Comment (yanns) [reply modules/core/src/main/scala/sangria/package.scala:10]: You can change the text how you like.
I'd just like that we don't mention any other libraries / projects to stay neutral.

File: modules/core/src/main/scala/sangria/parser/package.scala (hunk 7 above)
Comment (performantdata) [reply modules/core/src/main/scala/sangria/parser/package.scala:6]: Oh, like:
```scala
doesThisCompile {
  val mapper: sangria.parser.SourceMapper = new DefaultSourceMapper...
}
```
I haven't built a test like that yet.

I'm busy most of today, but can maybe get to it later.

**Comment by performantdata** [comment]
:shipit:

**Comment by performantdata** [comment]
How/when does the CHANGELOG get updated?

**Comment by performantdata** [comment]
This should be added to the milestone. @yanns

File: modules/core/src/main/scala/sangria/package.scala (hunk 6 above)
Comment (yanns) [reply modules/core/src/main/scala/sangria/package.scala:10]: Can you please change that?

**Review by yanns — CHANGES_REQUESTED** [review]
Great work!
Can you please remove the references to other libraries? Thanks!

**Comment by yanns** [comment]
> How/when does the CHANGELOG get updated?

manually, when we can.

File: modules/core/src/main/scala/sangria/package.scala (hunk 6 above)
Comment (performantdata) [reply modules/core/src/main/scala/sangria/package.scala]: > You can change the text how you like.

I thought you were giving me the choice.

**Review by yanns — APPROVED** [review]
💯

File: modules/core/src/main/scala/sangria/package.scala (hunk 6 above)
Comment (yanns) [reply modules/core/src/main/scala/sangria/package.scala]: You can change the text how you like *as long as* we don't mention any other libraries / projects to stay neutral. 😉

**Comment by yanns** [comment]
This change breaks some projects: https://github.com/sangria-graphql/sangria/issues/795
